components/sync15/src/telemetry.rs

```diff
@@ -0,0 +1,645 @@
+mod engine_tests {
+    use super::*;
+
+    #[test]
+    fn test_engine() {
+        let mut e = Engine::new("test_engine");
+        e.finished();
+        assert_json(&e, json!({"name": "test_engine", "when": 0.0}));
+    }
+
+    #[should_panic]
+    #[test]
+    fn test_engine_not_finished() {
+        let e = Engine::new("test_engine");
+        assert_json(&e, json!({}));
+    }
+
+    #[test]
+    fn test_incoming() {
+        let mut i = EngineIncoming::new();
+        i.applied(1);
+        i.failed(2);
+        let mut e = Engine::new("TestEngine");
+        e.incoming(i);
+        e.finished();
+        assert_json(
+            &e,
+            json!({"name": "TestEngine", "when": 0.0, "incoming": {"applied": 1, "failed": 2}}),
+        );
+    }
+
+    #[test]
+    fn test_outgoing() {
+        let mut o = EngineOutgoing::new();
+        o.sent(2);
+        o.failed(1);
+        let mut e = Engine::new("TestEngine");
+        e.outgoing(o);
+        e.finished();
+        assert_json(
+            &e,
+            json!({"name": "TestEngine", "when": 0.0, "outgoing": [{"sent": 2, "failed": 1}]}),
+        );
+    }
+
+    #[test]
+    fn test_failure() {
+        let mut e = Engine::new("TestEngine");
+        e.failure(SyncFailure::Http { code: 500 });
+        e.finished();
+        assert_json(
+            &e,
+            json!({"name": "TestEngine",
+             "when": 0.0,
+             "failureReason": {"name": "httperror", "code": 500}
+            }),
+        );
+    }
+
+    #[test]
+    fn test_raw() {
+        let mut e = Engine::new("TestEngine");
+        let mut inc = EngineIncoming::new();
+        inc.applied(10);
+        e.incoming(inc);
+        let mut out = EngineOutgoing::new();
+        out.sent(1);
+        e.outgoing(out);
+        e.failure(SyncFailure::Http { code: 500 });
+        e.finished();
+
+        assert_eq!(e.outgoing.len(), 1);
+        assert_eq!(e.incoming.as_ref().unwrap().applied, 10);
+        assert_eq!(e.outgoing[0].sent, 1);
+        assert!(e.failure.is_some());
+        serde_json::to_string(&e).expect("should get json");
+    }
+}
+
+//////////////////////////////////////////////////////////////////////////////
+//
+// A single sync. May have many engines, may have its own failure.
+//
+#[derive(Debug, Serialize)]
+pub struct Sync {
+    #[serde(flatten)]
+    when_took: Stopwatch,
+
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    engines: Vec<Engine>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(rename = "failureReason")]
+    failure: Option<SyncFailure>,
+}
+
+impl Sync {
+    pub fn new() -> Self {
+        Self {
+            when_took: Stopwatch::new(),
+            engines: Vec::new(),
+            failure: None,
+        }
+    }
+
+    pub fn engine(&mut self, mut e: Engine) {
+        e.finished();
+        self.engines.push(e);
+    }
+
+    pub fn failure(&mut self, failure: SyncFailure) {
+        assert!(self.failure.is_none());
```

Comment:
I'm not sure this assert is good. I could see a case where we end up calling `failure` and then lose error information (in the best case) as a result.

I suspect if it is called multiple times, the first error is the real error, and the rest of them are just caused by that. But maybe not.